cmd/orch-tester/orch_tester.go

```diff
@@ -549,3 +603,211 @@ func (s *streamerClient) getOrchestrators() ([]*orch, error) {
+	}
+
+	var rendSegs []*m3u8.MediaSegment
+	var sourceSegs []*m3u8.MediaSegment
+	for _, v := range renditionPl.Segments {
+		if v != nil {
+			rendSegs = append(rendSegs, v)
+		}
+	}
+	for _, v := range sourcePl.Segments {
+		if v != nil {
+			sourceSegs = append(sourceSegs, v)
+		}
+	}
+	if len(rendSegs) == 0 || len(sourceSegs) == 0 {
+		return nil, nil, "", fmt.Errorf("no segments found")
+	}
+
+	r := rand.Int()
+	segIdx := r % int(math.Min(float64(len(renditionPl.Segments)), float64(len(sourcePl.Segments))))
+	seqNo := sourcePl.Segments[segIdx].SeqId // Assumes that segIdx points to the same seqNo in the slice of sources and renditions
+	rendition, err = s.downloadSegment(fmt.Sprintf("%v/%v", s.broadcaster, rendSegs[seqNo].URI))
+	if err != nil {
+		return nil, nil, "", err
+	}
+
+	source, err = s.downloadSegment(fmt.Sprintf("%v/%v", s.broadcaster, sourceSegs[seqNo].URI))
+	if err != nil {
+		return nil, nil, "", err
+	}
+
+	return source, rendition, rendSegs[seqNo].URI, err
+}
+
+func (s *streamerClient) downloadSegment(url string) ([]byte, error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	res, err := s.client.Do(req)
+	if err != nil {
+		glog.Error(err)
+		return nil, err
+	}
+	body, err := ioutil.ReadAll(res.Body)
+	defer res.Body.Close()
+	if err != nil {
+		glog.Error(err)
+		return nil, err
+	}
+
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		return nil, fmt.Errorf("%v", string(body))
+	}
+	return body, nil
+}
+
+func (s *streamerClient) downloadMasterPlaylist(manifestID string) (*m3u8.MasterPlaylist, error) {
+	url := fmt.Sprintf("%v/stream/%v_0_0.m3u8", s.broadcaster, manifestID)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := s.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		return nil, errors.New(string(body))
+	}
+
+	mpl := m3u8.NewMasterPlaylist()
+	if err := mpl.Decode(*bytes.NewBuffer(body), true); err != nil {
+		return nil, err
+	}
+
+	return mpl, nil
+}
+
+func (s *streamerClient) downloadMediaPlaylist(uri string) (*m3u8.MediaPlaylist, error) {
+	url := fmt.Sprintf("%v/stream/%v", s.broadcaster, uri)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := s.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		return nil, errors.New(string(body))
+	}
+
+	pl, err := m3u8.NewMediaPlaylist(1, 1)
+	if err != nil {
+		return nil, err
+	}
+	if err := pl.Decode(*bytes.NewBuffer(body), true); err != nil {
+		return nil, err
+	}
+
+	for _, s := range pl.Segments {
+		if s == nil {
+			break
```

Comment:
Still don't understand what is the point of that loop?
Also, there shouldn't be nil sigments in the middle of slice - you see nil segments because whole slice is allocated at creation, but not necessarily filled when decoded.
If you need number of segments in the playlist you should call `pl.Len()`

Reply:
Not sure either, I think it lingered from earlier. Will remove.